Signal/src/util/OWSBackup.m

```diff
@@ -0,0 +1,908 @@
+
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        [self prepareForImport:srcZipPath];
+
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self complete];
+        });
+    });
+}
+
+- (void)prepareForImport:(NSString *)srcZipPath
+{
+    OWSAssert(srcZipPath.length > 0);
+
+    DDLogInfo(@"%@ %s.", self.logTag, __PRETTY_FUNCTION__);
+
+    NSString *documentDirectoryPath = OWSFileSystem.appDocumentDirectoryPath;
+    NSString *rootDirName = [OWSBackup_DirNamePrefix stringByAppendingString:[NSUUID UUID].UUIDString];
+    NSString *rootDirPath = [documentDirectoryPath stringByAppendingPathComponent:rootDirName];
+    NSString *backupDirPath = [rootDirPath stringByAppendingPathComponent:@"Contents"];
+    NSString *backupZipPath = [rootDirPath stringByAppendingPathComponent:srcZipPath.lastPathComponent];
+    self.backupDirPath = backupDirPath;
+    self.backupZipPath = backupZipPath;
+    DDLogInfo(@"%@ rootDirPath: %@", self.logTag, rootDirPath);
+    DDLogInfo(@"%@ backupDirPath: %@", self.logTag, backupDirPath);
+    DDLogInfo(@"%@ backupZipPath: %@", self.logTag, backupZipPath);
+
+    [OWSFileSystem ensureDirectoryExists:rootDirPath];
+    [OWSFileSystem protectFolderAtPath:rootDirPath];
+    [OWSFileSystem ensureDirectoryExists:backupDirPath];
+
+    NSError *error = nil;
+    BOOL success = [[NSFileManager defaultManager] copyItemAtPath:srcZipPath toPath:backupZipPath error:&error];
+    if (!success || error) {
+        OWSFail(@"%@ failed to copy backup zip: %@, %@", self.logTag, srcZipPath, error);
+        return [self fail];
+    }
+
+    if (self.isCancelledOrFailed) {
+        return;
+    }
+    if (![self unzipFilePath]) {
+        return [self fail];
+    }
+    if (self.isCancelledOrFailed) {
+        return;
+    }
+    if (![self extractEncryptionKey]) {
+        return [self fail];
+    }
+    if (self.isCancelledOrFailed) {
+        return;
+    }
+    if (![self isValidBackup]) {
+        return [self fail];
+    }
+    if (self.isCancelledOrFailed) {
+        return;
+    }
+    if (![self enqueueBackupRestore]) {
+        return [self fail];
+    }
+}
+
+- (BOOL)extractEncryptionKey
+{
+    DDLogInfo(@"%@ %s.", self.logTag, __PRETTY_FUNCTION__);
+
+    NSString *encryptionKeyFilePath =
+        [self.backupDirPath stringByAppendingPathComponent:OWSBackup_EncryptionKeyFilename];
+    if (![[NSFileManager defaultManager] fileExistsAtPath:encryptionKeyFilePath]) {
+        return NO;
+    }
+    NSData *_Nullable encryptionKeyData = [NSData dataWithContentsOfFile:encryptionKeyFilePath];
+    if (!encryptionKeyData) {
+        return NO;
+    }
+    OWSAES256Key *encryptionKey = [OWSAES256Key keyWithData:encryptionKeyData];
+    if (!encryptionKey) {
+        return NO;
+    }
+    self.encryptionKey = encryptionKey;
+
+    NSError *error = nil;
+    BOOL success = [[NSFileManager defaultManager] removeItemAtPath:encryptionKeyFilePath error:&error];
```

Comment:
There are hard problems here that we'll need to confront if we do user-facing backup.  For debugging purposes, the approach here should be sufficient.

Restoring a backup involves multiple steps and there's no way to do them atomically.  Therefore there's always going to be a risk that an incomplete restore leaves you with an invalid install.  Worst case is database + password mismatch.  

In terms of interrupted restores, we can err on the side of making a best effort to resume or backing off as soon as any serious errors are detected.  I've chosen to do the latter.

Circling back to your original question, your comment is in the "restore preparation" logic, not the more fraught "restore completion" logic.  If "backup preparation" is interrupted, no restore takes place and no harm is done.

Reply:
If/when we do user-facing backup, if we only support restores post-registration, the stakes aren't that high if a restore is interrupted and the user is left with an invalid install since by definition they can just re-register and re-restore.